### PR TITLE
Typed tables

### PR DESCRIPTION
**Pull request by danielraggi**

### CODE REVIEW

File: src/strategies/property_readers.sml [hunk 1]
```diff
@@ -137,6 +29,20 @@ fun label str = [str];
 fun collection str = if str = "NONE" then []
                      else map stringTrim (String.tokens (fn c => c = #",") str);
 
+(*)
```
Comment (aaronstockdill) [inline src/strategies/property_readers.sml]: Why is this commented out? If it is potentially useful in future, uncomment it but leave a comment as to why it is there. If it is a relic of old ideas, please delete it.

File: src/strategies/property_readers.sml [hunk 2]
```diff
@@ -169,8 +75,14 @@ fun dimension str =
     in
         List.foldr (fn (a, b) => a @ b) [] (dimensionsSplitOut @ dimensionsNoLabels)
     end;
+(*)
```
Comment (aaronstockdill) [inline src/strategies/property_readers.sml]: Why is this commented out? If it is potentially useful in future, uncomment it but leave a comment as to why it is there. If it is a relic of old ideas, please delete it.

File: src/util/type.sml [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+                                                 in if c = EQUAL then compare (u,u') else c
+                                                 end
+  | compare (Function _, _) = LESS
+  | compare (_, Function _) = GREATER
+  | compare (Constructor (s,t), Constructor (s',t')) = let val c = String.compare (s,s')
+                                                       in if c = EQUAL then compare (t,t') else c
+                                                       end
+  ;
+
+fun typeToString (Ground s) = s
```
Comment (aaronstockdill) [inline src/util/type.sml]: Please rename to `toString` for consistency.

File: src/util/type.sml [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+  | typeToString (Var s) = "'" ^ s
+  | typeToString (Pair (t,u)) = "(" ^ (typeToString t) ^ " * " ^ (typeToString u) ^ ")"
+  | typeToString (Function (t, (Function (u,v)))) =
+    (typeToString t) ^ " -> (" ^ (typeToString (Function (u,v))) ^ ")"
+  | typeToString (Function (t,u)) = (typeToString t) ^ " -> " ^ (typeToString u)
+  | typeToString (Constructor (s,t)) = (typeToString t) ^ " " ^ s;
+
+fun typeDebugString (Ground s) = "Ground \"" ^ s ^ "\""
```
Comment (aaronstockdill) [inline src/util/type.sml]: Consider renaming to `toDebugString` for consistency.

**Comment by danielraggi** [comment]
Types and attributes are read from tables. Properties are either Simple, Typed, or Attr. The idea is that tokens can be typed, and types can be given attributes (that every token with that type would have). From types we can potentially extract information like arity (e.g., is this an operator or a constant?), and the attributes should be things like "text" or "icon", or "emergent-symbol" which can be used to calculate cognitive properties.

Moreover, correspondences match properties not by string-equality but by a more sophisticated matching function, so that untyped or unattributed elements can match typed things, etc. There are also type variables which will match anything, so there's a unification-like matching (less sophisticated, because it takes all type variables as simple blanks or holes that can be filled with anything, so variable names don't matter).

File: src/strategies/property.sml [hunk 5]
```diff
@@ -10,19 +11,81 @@ sig
...
+  without the need to have type or attribute information, and type-matching when
+  there is type information *)
+fun propertyMatch (Simple s, p) = (s = nameOf p)
+  | propertyMatch (p, Simple s) = (s = nameOf p)
+  | propertyMatch (Typed (s,t), Typed (s',t')) = (s = s' andalso Type.match t t')
+  | propertyMatch (Attr (s,_), Attr (s',_)) = (s = s')
+  | propertyMatch _ = false
+
+fun toString (Simple s) = s
+  | toString (Typed (s,t)) = s ^ " : " ^ (Type.typeToString t)
```
Comment (aaronstockdill) [inline src/strategies/property.sml]: Change `Type.typeToString` to `Type.toString` (see comment in `type.sml`).

File: src/util/type.sml [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+  | typeDebugString (Var s) = "Var \"" ^ s ^ "\""
+  | typeDebugString (Pair (t,u)) = "Pair(" ^ (typeDebugString t) ^ ", "
+                                 ^ (typeDebugString u) ^ ")"
+  | typeDebugString (Function (t,u)) = "Function(" ^ (typeDebugString t) ^ ", "
+                                     ^ (typeDebugString u) ^ ")"
+  | typeDebugString (Constructor (s,t)) = "Constructor(\"" ^ s ^ "\", " ^ (typeDebugString t) ^ ")";
+
+fun vartype str =
```
Comment (aaronstockdill) [inline src/util/type.sml]: Rename to `fromString` for consistency.

File: src/strategies/property.sml [hunk 7]
```diff
@@ -10,19 +11,81 @@ sig
...
+                else let
+                          fun dropEnds [] = []
+                            | dropEnds [x] = []
+                            | dropEnds [x, y] = []
+                            | dropEnds (x::xs) = List.rev (List.tl (List.rev xs));
+                          val s' = String.implode (dropEnds (String.explode s));
+                      in
+                        Attr (r,map stringTrim (String.tokens (fn c => c = #";") s'))
+                      end
+            else Typed (r, Type.vartype s)
```
Comment (aaronstockdill) [inline src/strategies/property.sml]: Change `Type.vartype` to `Type.fromString` (see comment in `type.sml`).

File: src/util/type.sml [hunk 8]
```diff
@@ -0,0 +1,152 @@
+import "util.logging";
+
```
Comment (aaronstockdill) [inline src/util/type.sml:2]: Consider adding a signature so that the functions are documented clearly for others to use.

**Review by aaronstockdill — CHANGES_REQUESTED** [review]
Looks great! There are one or two minor things to tidy up (mostly names to improve consistency) then I'm happy to merge it.

File: src/strategies/property_readers.sml (hunk 1 above)
Comment (danielraggi) [reply src/strategies/property_readers.sml]: Done

File: src/strategies/property_readers.sml (hunk 2 above)
Comment (danielraggi) [reply src/strategies/property_readers.sml]: Done

File: src/util/type.sml (hunk 3 above)
Comment (danielraggi) [reply src/util/type.sml]: Done

**Review by aaronstockdill — APPROVED** [review]

**Pull request merged**